git: worktree, Eliminate duplicate Tree.FindEntry calls in checkout

cedric-appdirect opened the pull request:
During checkout operations, Tree.FindEntry was being called twice per file:
1. In checkoutChange() to obtain the TreeEntry
2. In checkoutChangeRegularFile() via Tree.File(), which internally calls FindEntry again with the same path

This optimization passes the TreeEntry directly from checkoutChange to checkoutChangeRegularFile, eliminating the redundant tree traversal. The blob is now retrieved directly using object.GetBlob() with the hash from the already-found TreeEntry.

Performance impact measured with BenchmarkAlternatesPerformance (40,000 files):
- Tree.FindEntry time reduced from 120.45s to 62.45s (~52% reduction)
- Savings: ~1.55ms per file checkout operation
- Tree.File() usage dropped from 62s+ to negligible (10ms, used elsewhere)

The change maintains identical behavior while significantly improving performance for large checkouts, particularly beneficial for repositories using alternates where object lookups are more expensive.

cedric-appdirect committed:
git: worktree, Eliminate duplicate Tree.FindEntry calls in checkout

During checkout operations, Tree.FindEntry was being called twice per file:
1. In checkoutChange() to obtain the TreeEntry
2. In checkoutChangeRegularFile() via Tree.File(), which internally calls
   FindEntry again with the same path

This optimization passes the TreeEntry directly from checkoutChange to
checkoutChangeRegularFile, eliminating the redundant tree traversal. The
blob is now retrieved directly using object.GetBlob() with the hash from
the already-found TreeEntry.

Performance impact measured with BenchmarkAlternatesPerformance (40,000 files):
- Tree.FindEntry time reduced from 120.45s to 62.45s (~52% reduction)
- Savings: ~1.55ms per file checkout operation
- Tree.File() usage dropped from 62s+ to negligible (10ms, used elsewhere)

The change maintains identical behavior while significantly improving
performance for large checkouts, particularly beneficial for repositories
using alternates where object lookups are more expensive.

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

onee-only commented on plumbing/object/tree.go:
```diff
@@ -89,6 +89,21 @@ func (t *Tree) File(path string) (*File, error) {
+	}
+
+	return NewFile(path, e.Mode, blob), nil
+}
```
`Tree.TreeEntryFile()` does the same thing as this. However, the naming `FileFromEntry` and the error handling seem like good improvements. What are your thoughts on removing `Tree.TreeEntryFile()` and using this instead?

Also, as in `Tree.TreeEntryFile()`, taking only `e *TreeEntry` should be sufficient. The `path` parameter is redundant since it's equivalent to `e.Name`.

https://github.com/go-git/go-git/blob/5d40cdde8ca0df28c69852d2082df0d420cd2436/plumbing/object/tree.go#L119-L127

onee-only replied:
You're right. I thought `TreeEntry.Name` had the full path. 😅

cedric-appdirect committed:
plumbing: object, Use FileFromEntry in Change.Files for correct path

Change.Files() was using TreeEntryFile() which sets File.Name to the
TreeEntry.Name (basename only, e.g., "main.go"). This is inconsistent
because ChangeEntry.Name already contains the full path (e.g.,
"examples/clone/main.go").

This change updates Change.Files() to use FileFromEntry() instead,
which accepts the full path as a parameter and sets File.Name correctly.

Impact:
- File.Name returned by Change.Files() now contains the full path
  instead of just the basename
- Error messages change from "object not found" to "file not found"
  when the blob cannot be found, providing better context
- The file mode check is now performed before attempting to fetch
  the blob, avoiding unnecessary work for non-file entries

This makes the behavior consistent with Tree.File() which also returns
a File with the full path in its Name field.

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
plumbing: object, Remove unused TreeEntryFile method

TreeEntryFile has been replaced by FileFromEntry which provides the
same functionality but with correct full path handling. The previous
commit migrated the last usage in Change.Files() to FileFromEntry.

TreeEntryFile only set File.Name to the basename (TreeEntry.Name),
which was inconsistent with other methods like Tree.File() that use
the full path. FileFromEntry addresses this by accepting the path as
a parameter, making the API more explicit about what path will be
used for the returned File.

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, Add benchmark for checkout operations

Add BenchmarkCheckout with three sub-benchmarks to measure checkout
performance:

- SameTree: Benchmarks creating new branches from current HEAD with
  the same tree. This is the common case that benefits from fast path
  optimization when the tree hasn't changed.

- SameTreeForce: Same as above but with Force flag, to ensure no
  regression in the force checkout path.

- DifferentBranch: Benchmarks switching between different branches to
  measure the slow path where tree differences must be computed.

These benchmarks help verify that checkout optimizations (such as
eliminating redundant FindEntry calls) don't cause regressions and
provide measurable improvements.

Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
storage: filesystem, Fix findInAlternates false success on all-miss path

findInAlternates returned (zero, nil) instead of ErrObjectNotFound
when every alternate returned ErrObjectNotFound. The concurrent
goroutines swallowed those errors to nil, so g.Wait() returned nil
with found still false, producing a false success for callers.

Also add a nil guard to object.Tree.FileFromEntry so that a nil
*TreeEntry returns ErrFileNotFound instead of panicking, consistent
with the existing File(path) method.

Made-with: Cursor
Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>

cedric-appdirect committed:
git: worktree, update go-git-fixtures import from v5 to v6. Fixes #1748

Update the import in worktree_bench_test.go from go-git-fixtures/v5 to v6,
which is already used consistently throughout the test suite (64 other files).
In fixtures v6, DotGit() now returns (fs, error), so add error handling to
match the API change.

This resolves the CI failures on PR #1748 caused by upstream main migrating
to fixtures v6.

Assisted-by: OpenCode with Claude Opus 4.6 <noreply@anthropic.com>
Signed-off-by: Cedric BAIL <cedric.bail@appdirect.com>
